### DISCUSSION TITLE
Prevent SQLAlchemy from Caching Objects in Multi-layered Views Without a Primary Key

### DISCUSSION THREAD

**Discussion by ManiMozaffar** [sqlalchemy/sqlalchemy#11968] [Usage Questions]
Following this guideline, I built a view hierachy where a view was created by querying another view
https://github.com/sqlalchemy/sqlalchemy/wiki/Views

I'm using SQLAlchemy views, and when I query multi-layered views, I'm getting inconsistent results. The problem is that without a primary key, SQLAlchemy caches objects incorrectly. The database returns the correct rows, but SQLAlchemy repeats the first row for all results, like this:

```python
cursor = [1,2,3]
session = [1,1,1]
```

From my understanding, SQLAlchemy relies on primary key to use it as a unique identifier for caching key.
For my case, inconsistency happens because there's no primary key, so SQLAlchemy can't uniquely identify the rows, causing it to reuse the first object in the cache. I don't want to define a primary key for the view because:

- Setting all columns as the primary key is inefficient and consumes too much memory.
- Adding a UUID as a primary key feels unnecessary and confusing on the database interface.

What I really want is to prevent SQLAlchemy from caching objects from this view.

**Comment by CaselIT** [comment]
Hi,

It's not clear what you are doing.
When using core sqlalchemy does no not need primary keys, while when using the orm a way of uniquely identifying each orm instance is required.

What you describe seems like what the orm does, if you don't have a way of uniquely identifying the rows then you cannot use it, use the core directly.

For more suggestion you will need to provide some examples

**Reply by ManiMozaffar** [reply]
Yes I'm using ORM. I'll provide example, using core could be a solution but the problem is that core return data unstructured and unmapped so I need to clean them myself which is a bit annoying, for example a UUID is not UUID but asyncpg UUID. I still like to have the "type conversion" thing, without caching and attaching and other stuff that the session(ORM) does for me.
So you're saying there's no way to skip uniquely identifying the row? because my row cannot be uniquely identified, it's a view's row, it changes everytime.

**Reply by gordthompson** [reply]
re: "caching … that the session(ORM) does for me"

https://docs.sqlalchemy.org/en/20/orm/session_basics.html#is-the-session-a-cache

re: "my row cannot be uniquely identified, it's a view's row, it changes everytime"

The rows returned may change each time the view is invoked, but in order to use ORM there must be a way to uniquely identify each row that is returned. Trying to use an ORM object for a table (or view) with no primary key results in

> sqlalchemy.exc.ArgumentError: Mapper Mapper[Thing(thing)] could not assemble any primary key columns for mapped table 'thing'

**Reply by CaselIT** [reply]
A few things:
if you are using ORM you must be specifying either a primary key in the view or as a mapper arg. If you provide no key an error is raised, like mentioned by gord

> for example a UUID is not UUID but asyncpg UUID

  I don't really follow what you mean here. Any type conversion logic happens in the core, the orm instantiates orm classes, it does not convert the single items

> So you're saying there's no way to skip uniquely identifying the row? 

The orm needs to uniquely identify the row returned by a single select. 

> because my row cannot be uniquely identified, it's a view's row, it changes everytime.

It seems strange. So you have multiple rows with the same values? if that's something that's expected then ORM cannot really be used here without finding a way of discriminating each row

**Comment by ManiMozaffar** [comment]
@CaselIT I tried reproducing the issue, took me sometime. Figured out root of issue, I think this is a SQLA bug. or at least this behavior should be documented if it's by-design.

> if you are using ORM you must be specifying either a primary key in the view or as a mapper arg. If you provide no key an error is raised, like mentioned by gord

I think that's not entirely true, from my understanding it actually tries to find the primary key somehow.  Look at the code snippet that I've sent in below, if you comment out "author_id" on the view model,  you get the exception that your model doesn't have any primary key.
The bug happened because ORM tried to smartly identify the primary key, which "author id" was identified, but the issue is that the actual primary key is composite of "author id & id" for some reasons.

So given for "same authors", yes I always get same row. Look at the image:
<img width="1440" alt="Screenshot 2024-10-09 at 12 00 05" src="https://github.com/user-attachments/assets/b78fc4c0-a1f2-4749-b6e9-ee9e16aad3e8">

Name: SQLAlchemy
Version: 2.0.28

Code snippet:
```python
import typing
from contextlib import asynccontextmanager
from datetime import UTC, datetime, timedelta
from typing import Any, Self
from uuid import uuid4

import sqlalchemy as sa
from rich import print
from sqlalchemy.exc import NoInspectionAvailable
from sqlalchemy.ext import compiler
from sqlalchemy.ext.asyncio import async_sessionmaker, create_async_engine
from sqlalchemy.orm import DeclarativeBase, Mapped, MappedAsDataclass, mapped_column
from sqlalchemy.schema import DDLElement


class BaseSQL(MappedAsDataclass, DeclarativeBase):
    def __init_subclass__(cls, *args, **kw) -> None:
        if hasattr(cls, "__table__"):
            table_clause = cls.__table__  # type: ignore
            if isinstance(table_clause, View):
                for k in cls.__annotations__:
                    annotation = cls.__annotations__[k]
                    annotation = typing.get_args(annotation)[0]  # remove Mapped
                    cls.__annotations__[k] = annotation

        super().__init_subclass__(*args, **kw)


class CreateView(DDLElement):
    def __init__(self, name: str, selectable: sa.Select):
        self.name = name
        self.selectable = selectable


class DropView(DDLElement):
    def __init__(self, name, cascade=False, if_exists=False):
        self.name = name
        self.cascade = cascade
        self.if_exists = if_exists


@compiler.compiles(CreateView)
def _create_view(element: CreateView, compiler, **kw):
    return 'CREATE VIEW "%s" AS %s' % (
        element.name,
        compiler.sql_compiler.process(element.selectable, literal_binds=True),
    )


@compiler.compiles(DropView)
def _drop_view(element: DropView, compiler, **kw):
    text = "DROP VIEW "
    if element.if_exists:
        text += "IF EXISTS "
    text += f'"{element.name}"'
    if element.cascade:
        text += " CASCADE"
    return text


def view_exists(ddl, target, connection, **kw):
    try:
        return ddl.name in sa.inspect(connection).get_view_names()
    except NoInspectionAvailable:
        return False


def view_doesnt_exist(ddl, target, connection, **kw):
    return not view_exists(ddl, target, connection, **kw)


class View(sa.TableClause):
    exclude_in_sqlite: bool = False
    inherit_cache: bool = False

    @classmethod
    def from_name(cls, name: str, *columns: sa.ColumnClause[Any], **kw: Any) -> Self:
        return cls(name, *columns, **kw)


def view(name, metadata, selectable, *, cascade=True):
    t = View.from_name(name)
    t._columns._populate_separate_keys(col._make_proxy(t) for col in selectable.selected_columns)
    sa.event.listen(
        metadata,
        "after_create",
        CreateView(name, selectable).execute_if(callable_=view_doesnt_exist),  # type: ignore
    )
    sa.event.listen(
        metadata,
        "before_drop",
        DropView(name, cascade=cascade).execute_if(callable_=view_exists),  # type: ignore
    )
    return t


class Event(BaseSQL):
    __tablename__ = "Events"
    event: Mapped[str] = mapped_column(sa.String())
    created_at: Mapped[datetime] = mapped_column(sa.DateTime(timezone=True))
    id: Mapped[str] = mapped_column(primary_key=True, default_factory=lambda: str(uuid4()))
    author_id: Mapped[str] = mapped_column(primary_key=True, default_factory=lambda: str(uuid4()))


def event_period_view_query():
    event_transitions = sa.select(
        Event.author_id,
        Event.event,
        Event.created_at,
        sa.func.lead(Event.event)
        .over(
            partition_by=Event.author_id,
            order_by=Event.created_at,
        )
        .label("next_event"),
        sa.func.lead(Event.created_at)
        .over(
            partition_by=Event.author_id,
            order_by=Event.created_at,
        )
        .label("next_created_at"),
    ).subquery()

    event_periods = sa.select(
        event_transitions.c.author_id,
        event_transitions.c.event,
        event_transitions.c.created_at.label("start_datetime"),
        sa.func.coalesce(event_transitions.c.next_created_at, sa.func.now()).label("end_datetime"),
    ).select_from(event_transitions)
    return event_periods


class EventPeriod(BaseSQL):
    __tablename__ = "EventPeriods"
    __table__ = view(__tablename__, BaseSQL.metadata, event_period_view_query())
    author_id: Mapped[str]
    event: Mapped[str]
    start_datetime: Mapped[datetime]
    end_datetime: Mapped[datetime]


@asynccontextmanager
async def init_db():
    engine = create_async_engine("sqlite+aiosqlite:///:memory:")
    db = async_sessionmaker(engine, expire_on_commit=False, autobegin=False)

    async with engine.begin() as conn:
        await conn.run_sync(BaseSQL.metadata.create_all)

    yield db


async def test():
    query = sa.select(EventPeriod)
    async with init_db() as db:
        async with db.begin() as session:
            now = datetime.now(UTC)
            author_id = str(uuid4())
            session.add_all(
                [
                    Event(author_id=author_id, event="first", created_at=now - timedelta(days=3)),
                    Event(author_id=author_id, event="second", created_at=now - timedelta(days=2)),
                    Event(author_id=author_id, event="third", created_at=now - timedelta(days=1)),
                ]
            )
            await session.flush()

            result = (await session.scalars(query)).all()
            print("Result from `AsyncSession`:")
            print(50 * "--")
            print(result)

            core_conn = await session.connection()
            result = (await core_conn.execute(query)).all()
            print("Result from `AsyncConnection`:")
            print(50 * "--")

            print(result)


if __name__ == "__main__":
    import asyncio

    asyncio.run(test())
```

**Reply by ManiMozaffar** [reply]
let me know if I should file this as a bug @CaselIT

**Reply by CaselIT** [reply]
There doesn't seem any bug here.

It's working as it have been explained 

Also not entirely sure why you are doing 
 ```py
annotation = typing.get_args(annotation)[0]  # remove Mapped
cls.__annotations__[k] = annotation

**Reply by ManiMozaffar** [reply]
> It's working as it have been explained

I didn't find anywhere in documentation that this was explained, maybe I'm wrong. 
It picked only one "unique identifier"  from unique composition, which lead to that issue. 
I think ideally it should have raised exception saying that the table must explicitly have identifier. Maybe that's too hard I don't know just an idea as a user.

> Also not entirely sure why you are doing

Because otherwise it fails on initialization of the model.

```traceback
Traceback (most recent call last):
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/example.py", line 135, in <module>
    class EventPeriod(BaseSQL):
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_api.py", line 617, in __init_subclass__
    super().__init_subclass__(**kw)
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_api.py", line 836, in __init_subclass__
    _as_declarative(cls._sa_registry, cls, cls.__dict__)
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 244, in _as_declarative
    return _MapperConfig.setup_mapping(registry, cls, dict_, None, {})
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 325, in setup_mapping
    return _ClassScanMapperConfig(
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 563, in __init__
    self._setup_dataclasses_transforms()
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/decl_base.py", line 1087, in _setup_dataclasses_transforms
    _AttributeOptions._get_arguments_for_make_dataclass(
  File "/Users/manimozaffar/Desktop/projects/noah/noah-server/.venv/lib/python3.12/site-packages/sqlalchemy/orm/interfaces.py", line 289, in _get_arguments_for_make_dataclass
    assert False, "Mapped[] received without a mapping declaration"
           ^^^^^
AssertionError: Mapped[] received without a mapping declaration
```

**Reply by CaselIT** [reply]
> It picked only one "unique identifier" from unique composition, which lead to that issue

Your select has a only one of the pk column, so that's what the orm has ti work with.

In your case since you are using a select as the table it will likely try to figure it out. If it's mentioned it will be in the Mapper.primary_key docs

> Because otherwise it fails on initialization of the model.

Interesting. It may be that using dataclasses with a table provided is not tested

**Reply by ManiMozaffar** [reply]
> If it's mentioned it will be in the Mapper.primary_key docs

I already read that, the documentation is saying that "MOST" orms requires to construct a class where the primary keys aren't set explicitly, which is a good behavior IMO.

> In your case since you are using a select as the table it will likely try to figure it out

That's the problem, it's trying to figure out, but then resolves to a primary key which is not true, therefore the result from ORM looks way off. 

> Interesting. It may be that using dataclasses with a table provided is not tested

Yes I also think that's the root of issue.


I think for my own solution, I'll go with a synthetic ID (which would be a uuid generator on DB side) and always use that as cache key/primary key **on the VIEW** objects.

**Reply by CaselIT** [reply]
> I think for my own solution, I'll go with a synthetic ID (which would be a uuid generator on DB side) and always use that as cache key/primary key on the VIEW objects.

you can also probably add a row_number in your view and use that as part of the primary key.

> That's the problem, it's trying to figure out, but then resolves to a primary key which is not true, therefore the result from ORM looks way off.

I guess in this case it's a bit unfortunate.
```
import sqlalchemy as sa

t = sa.Table('x', sa.MetaData(), sa.Column('id', sa.Integer, primary_key=True), sa.Column('name', sa.Integer, primary_key=True))
print(t.primary_key) # prints both keys.
q = select(t.c.id).subquery()
print(q.primary_key) # prints only id
```
@zzzeek do you think it would make sense to change how this works?
I'm thinking something like populating the pk of the subquery only if all the columns of the original table are in the select?

> Yes I also think that's the root of issue.

I may open an issue here

**Reply by CaselIT** [reply]
> I may open an issue here

#11973

**Comment by ManiMozaffar** [comment]
I ended up writing my [own Session](https://github.com/ManiMozaffar/naked-sqla) where no object tracking happens -> simply engine +object mapping. 

The identity key logic in ORM seems to be buggy, check this [failing test case](https://github.com/ManiMozaffar/naked-sqla/blob/main/tests/test_complicated_update.py) where I'm updating event to value=2, but ORM is returning me value=1 where engine/core would return me value=2 (this time no view, simply some normal select/update statement)

**Reply by CaselIT** [reply]
You probably need to add populate_existing.
Reading the docs will give you a better understanding of why the session does what it does. 

In any case if that works for you, then great!

**Reply by ManiMozaffar** [reply]
Yes I'm aware how populate_existing works, but that has nothing to do with this bug.

Here's the proof:

<img width="549" alt="Screenshot 2024-10-12 at 18 42 52" src="https://github.com/user-attachments/assets/a7df7c41-63c8-4623-894b-828b7447756f">
<img width="1441" alt="Screenshot 2024-10-12 at 18 43 04" src="https://github.com/user-attachments/assets/a31e5a9f-46f3-4c08-bc95-aa0228754cb2">



Now if I replace the subquery with this query that I wrote, it passes. (also removing the populate_existing)

<img width="589" alt="Screenshot 2024-10-12 at 19 03 12" src="https://github.com/user-attachments/assets/6801a9ea-623f-433c-8e93-1d40a1b411b3">


> This operation in either form indicates that objects being returned from a query should be unconditionally re-populated from their contents in the database

Well you can also argue that I can expunge my object from session, run that faulty query, and it should work. but what I'm really interested in, is to see why this is failing? Because that's a workaround, not a solution. I cannot assert where this problem might happen, and where it might not be, it's hard/impossible to tell without running the code.

But what I'm just curious what is the root of issue? Why changing the select statement (although it is still updating same row) change how the ORM behave? I could not find any documentation related to this. When I went through the code, I saw there are difference in how an object is loaded base on statement, but in this case it's always "dml.Insert".


Even though I set populate existing to true, but on the function where it initiate object from row, the populate existing is still false. maybe there's something wrong with what I'm doing.
after populating row with instance, the state obviously became correct again. 
<img width="1336" alt="Screenshot 2024-10-12 at 19 17 51" src="https://github.com/user-attachments/assets/c32385e1-ad39-4771-a9b4-f1cb6fafb97c">

**Reply by CaselIT** [reply]
> Yes I'm aware how populate_existing works, but that has nothing to do with this bug.

I don't know what bug you are referring to; screenshots of code snippet are really not much to work on.
If you have a reproduction case with a runnable script the best option is to open an issue

**Reply by ManiMozaffar** [reply]
Created @CaselIT  https://github.com/sqlalchemy/sqlalchemy/issues/11997

**Reply by CaselIT** [reply]
Thanks! So that can be properly tracked
